--- 
title: "Sphaerularia-Bombus infection RNA-Seq study"
output: plot_pca.html
author: Joe Colgan (joscolgan)
--- 

## Introduction:
This script reads in abundance (.h5) files generated by kallisto and calculates gene-level counts using tximport. A DESeq2 object is contructed and the gene-level counts are normalized. Using these normalised counts, a principal component analysis is performed.
This scripy output a .png image plot of the first two principal components.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport",
               "DESeq2",
               "ggplot2",
               "ReportingTools",
               "vsn",
               "RColorBrewer")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
## Create directory:
dir.create("results")
```

After loading libraries, the second step of the analysis involves reading in data:

```{r, message=FALSE} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$kallisto_output <- "./input/"
# Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output,
                                                     recursive = TRUE),
                                      pattern = ".h5",
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)
## Generate a dataframe for tximport containing sample name and 
## treatment:
## Assign the number of biological replicates
diapause_replicates <- 6
postdiapause_replicates <- 14
timepoint <- c(rep("diapause", diapause_replicates),
               rep("postdiapause", postdiapause_replicates))
treatment <- c(rep("infected", 3),
               rep("control", 3),
               rep("infected", 1),
               rep("control", 5),
               rep("infected", 6),
               rep("control", 2))
## Generate simple information dataframe:
samples <- as.data.frame(timepoint, paths$kallisto_files)
samples$treatment <- treatment
## Relevel by 'control' samples to ensure infected samples are compared against control
## rather than the reverse:
samples$timepoint <- relevel(x = samples$timepoint,
                             ref = "diapause")
# Read in file corresponding to transcripts to gene ids
tx_gene <- "input/rna_and_corresponding_gene_ids.txt"
tx2gene <- read.table(file = tx_gene,
                      header = FALSE,
                      col.names = c("TXNAME",
                                    "GENEID"))
# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files,
                       type    = "kallisto",
                       tx2gene = tx2gene,
                       countsFromAbundance = "no")

## Construct DESeq data set from txi object and sample information:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts,
                                      colData = samples,
                                      design  = ~ timepoint + treatment)
```


Generate plot using principal components calculated for transcriptomes from all queens:

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_txi,
           blind = FALSE)

## Extract first and second principal components:
pca_data    <- plotPCA(vsd,
                       intgroup = c("timepoint",
                                    "treatment"),
                       returnData = TRUE)
## Extract percentages for first two principal components:
percentvar <- round(100 * attr(pca_data, "percentVar"))
## Plot PCA:
plot <- ggplot(pca_data, aes(PC1, PC2,
                     color = timepoint)) +
        geom_point(aes(fill = timepoint,
                       shape = treatment),
                   size = 5) +
        xlab(paste0("PC1: ", percentvar[1], "% variance")) +
        ylab(paste0("PC2: ", percentvar[2], "% variance")) +
        coord_fixed() +
        scale_color_manual(values = c("blue", "grey", "black")) +
        theme_bw() +
        theme(axis.title = element_text(family = "Arial",
                                        color = "#666666",
                                        face = "bold",
                                        size = 16)) +
        theme(legend.position = "top") +
        theme(legend.text = element_text(colour = "black",
                                         size = 14,
                                         face = "bold"),
              legend.title = element_text(colour = "black",
                                          size = 14,
                                          face = "bold"))

## Save image to file:
ggsave(file = "./results/pca_plot.png",
       height = 5,
       width = 10)
```

Run lintr:
```{r, message = FALSE}
lintr::lint(file = "./plot_pca.Rmd")
```
