--- 
title: "Sphaerularia-Bombus infection RNA-Seq study"
output: differential_expression_analysis_diapause.html
author: Joe Colgan (joscolgan)
--- 

# Introduction
The purpose of this script is to read in transcript abundance files generated by the pseudoaligner Kallisto and 
perform differential expression analysis using the R package, DESeq2.
This script is a modification of scripts created by and freely available for reuse at:
https://github.com/wurmlab/Bter_neonicotinoid_exposure_experiment
The paper detailing the findings should be cited when scripts are reused:
Colgan, T.J., Fletcher, I.K., Arce, A.N., Gill, R.J., Ramos Rodrigues, A., Stolle, E., Chittka, L. and Wurm, Y.
Caste- and pesticide-specific effects of neonicotinoid pesticide exposure on gene expression in bumblebees. Molecular Ecology.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport",
               "DESeq2",
               "ggplot2",
               "ReportingTools",
               "vsn",
               "RColorBrewer")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
## Create directory:
dir.create("results")
```

After loading libraries, the second step of the analysis involves reading in data:

```{r, message=FALSE} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$kallisto_output <- "./input/"

# Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output,
                                                     recursive = TRUE),
                                      pattern = ".h5",
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

## Generate a dataframe for tximport containing sample name and 
## treatment:
## Assign the number of biological replicates
replicates <- 3
treatment <- c(rep("infected", replicates),
               rep("control", replicates))

## Generate simple information dataframe:
samples <- as.data.frame(treatment, paths$kallisto_files)

## Relevel by 'control' samples to ensure infected samples are compared against control
## rather than the reverse:
samples$treatment <- relevel(x = samples$treatment,
                             ref = "control")

# Read in file corresponding to transcripts to gene ids
tx_gene <- "input/rna_and_corresponding_gene_ids.txt"
tx2gene <- read.table(file = tx_gene,
                      header = FALSE,
                      col.names = c("TXNAME",
                                    "GENEID"))

# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files,
                       type    = "kallisto",
                       tx2gene = tx2gene,
                       countsFromAbundance = "no")

# Save object for use in other scripts
save(txi_counts, file = "diapause_raw_counts.Rdata")

## Construct DESeq data set from txi object and sample information:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts,
                                      colData = samples,
                                      design  = ~ treatment)

## Perform a pairwise comparison between treatments:
deseq_object  <- DESeq(deseq_txi,
                       test = "Wald",
                       betaPrior = TRUE)

## Extract results for infected vs. control:
deseq_results_infected <- results(deseq_object,
                                  contrast = c("treatment",
                                               "infected",
                                               "control"))

## Save object for use in generating heatmap:
save(deseq_object,
     file = "diapause_deseq_object.Rdata")

## Order results table by smallest raw pvalue:
deseq_results_infected_ordered <- deseq_results_infected[order(deseq_results_infected$pvalue), ]

## Subset genes by adjusted pvalue:
deseq_results_infected_ordered_sig <- subset(deseq_results_infected_ordered,
                                             padj < 0.05)

## Convert into a dataframe:
deseq_results_infected_ordered_sig_df <- as.data.frame(deseq_results_infected_ordered_sig)

## Write to file:
write.table(deseq_results_infected_ordered_significant_df,
            file = "results/de_genes_diapause.txt",
            sep = "\t",
            quote = FALSE)

# Save significant genes for heatmap input:
de_genes_infected <- rownames(deseq_results_infected_ordered_significant_df)

## Save objected containing significant differentially expressed genes:
save(de_genes_infected, file = "results/de_genes_diapause_infected.RData")
```

A downstream analysis is to explore gene ontology terms enriched within differentially expressed genes. 
The next step is to output the name of each gene, as well as raw p-values, which will be used within a rank-based test to test for enrichment of gene ontology terms within differentially expressed genes. 

```{r}
# Extract list of genes from the DESeq objects
# i) For S. bombi infected samples:
de_genes_for_topgo_infected <- data.frame(locus = row.names(deseq_results_infected),
                                          raw_pvalue = deseq_results_infected$pvalue)

# Order by p-values 
de_genes_for_topgo_infected <- de_genes_for_topgo_infected[order(de_genes_for_topgo_infected$raw_pvalue), ]

## Write to output:
write.table(de_genes_for_topgo_infected,
            file = "results/de_genes_for_topGO_diapause.txt",
            row.names = FALSE,
            sep = "\t",
            quote = FALSE)
```

Generate plot using principal components calculated for transcriptomes from all queens:

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object,
           blind = FALSE)

## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))

## Extract first and second principal components:
pca_data    <- plotPCA(vsd,
                       intgroup = c("treatment"),
                       returnData = TRUE)

## Extract percentages for first two principal components:
percentvar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2,
                     color = treatment,
                     shape = treatment)) +
        geom_point(aes(fill = treatment),
                   colour = "black", pch = 21, size = 5) +
        xlab(paste0("PC1: ", percentvar[1], "% variance")) +
        ylab(paste0("PC2: ", percentvar[2], "% variance")) +
        coord_fixed() +
        scale_color_manual(values = c("blue", "grey", "black")) +
        theme_bw() +
        theme(axis.title = element_text(family = "Arial",
                                        color = "#666666",
                                        face = "bold",
                                        size = 16)) +
        theme(legend.position = "top") +
        theme(legend.text = element_text(colour = "black",
                                         size = 14,
                                         face = "bold"),
              legend.title = element_text(colour = "black",
                                          size = 14,
                                          face = "bold"))

## Save file:
ggsave(file = "results/pca_output_diapause_samples.png",
       height = 10,
       width = 10)
```

Lastly, use lintr to check any potential style issues: 

```{r, message = FALSE}
## Run lintr:
lintr::lint("./sphaerularia_gene_expression_diapause.Rmd")
```


