--- 
title: "Sphaerularia-Bombus infection RNA-Seq study"
output: plot_combined_DEGs_heatmap.html
author: Joe Colgan (joscolgan)
--- 

## Introduction:

This script is for exploratory analysis of transcript quantification data and results of differential gene expression analysis.
Input files should be R objects saved and generated from previous scripts of DE analysis, including the object resulting from using tximport() on kallisto files, the significant transcript abundances selected from this object and a results table of significant genes, generated using DESeq2. 
This script will organise input data for generating heatmaps of significantly differentially expressed genes, based on statistical significance and log fold changes in expression. This script produces a heatmap that can be exported. 

```{r, message=FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2", "gplots", "heatmap3",
               "ggfortify", "rentrez",
               "reshape2", "scales", "ggpubr",
               "grid", "XML")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}

## Assign path for output:
output <- "results/heatmap_figure"
dir.create(output, recursive = TRUE)
```

Generate a combined heatmap:

```{r, message = FALSE}
load("./diapause_plot.Rdata")
load("./pd_plot.Rdata")

## Generate a combined plot:
ggarrange(diapause_plot,
          pd_plot,
          ncol = 1,
          nrow = 2,
          align = "hv",
          labels = c("A", "B"),
          heights = c(1.0, 2.5))

## Save output:
ggsave(paste(output,
             "/combined_heatmap.png",
             sep = ""),
       width = 20,
       height = 18)
```

Run lintr

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "./plot_combined_DEGs_heatmap.Rmd")
```


