---
title: "Nematode alters bumblebee gene expression"
output: compare_estimated_counts.html
author: Joe Colgan 
---

## Introduction:  
This script was used within the analysis of gene expression differences between uninfected bumblebee (Bombus terrestris) queens and Sphaerularia bombi infected queens. The script takes input estimated counts generated as output by the pseudoaligner, [Kallisto](https://pachterlab.github.io/kallisto/), calculates the gene-level estimated counts using [Tximport](http://bioconductor.org/packages/release/bioc/html/tximport.html) and compare gene-level counts for bumblebees collected from two time-points: during and after host diapause. 

The first step involves loading of libraries required for the analysis:
```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2",
               "ggplot2", "ggpubr", "rhdf5",
               "plyr", "ggfortify", "biomaRt",
               "vsn", "RColorBrewer", "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```
## Step One: Input results from kallisto
We use raw, non-normalised estimated counts.
Read in counts for diapause samples:

```{r, message=FALSE} 
## Set paths to folders containing output files from kallisto
path <- list()
path$sample_batch_info <- "./input/combined/sample_list.txt"
path$kallisto_output <- "./input/combined"
# Set relative paths:
path$kallisto_files_relative <- grep(x = list.files(path$kallisto_output,
                                                recursive = TRUE),
                                      pattern = "h5",
                                      value   = TRUE)
path$kallisto_files <- file.path(path$kallisto_output,
                                  path$kallisto_files_relative)
# Automatically extract file names from kallisto output for colnames
names(path$kallisto_files) <- gsub(path$kallisto_files_relative,
                                    pattern = "s
                                   /.*", replacement = "")
for (filenumber in 1:length(path$kallisto_files)) {
  current_name <- names(path$kallisto_files)[filenumber]
  current_file <- path$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples     <- data.frame(sample_name = names(path$kallisto_files))
## Read in sample information for room and batch:
samples_information <- read.table(file = path$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "timepoint",
                                                "status"),
                                  row.names = 1)

## Make sure uninfected ('control') is the reference:
samples_information$status <- relevel(x = samples_information$status,
                                         ref = "control")
```

## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
tx2gene <- read.table(file      = "input/combined/rna_and_corresponding_gene_ids.txt",
                      header    = FALSE,
                      col.names = c("TXNAME", "GENEID")) # i.e., transcript name, gene name

# Use tximport on kallisto files
## Counts are estimated counts from kallisto:
txi_counts <- tximport(path$kallisto_files,
                       type    = "kallisto",
                       tx2gene = tx2gene,
                       countsFromAbundance = "no")

## Count the number of genes per input data.frame:
nrow(txi_counts$counts)
```
## Step Four: Check the mean counts per timepoint:

```{r, message = FALSE}
## Subset the diapause samples:
txi_counts_diapause <- txi_counts$counts[, 15:20]
txi_counts_postdiapause <- txi_counts$counts[, 1:14]

## Calculate means for each time point:
means_df <- as.data.frame(cbind(rowMeans(txi_counts_diapause),
                                rowMeans(txi_counts_postdiapause)))
## Update column names:
colnames(means_df) <- c("Diapause", "Postdiapause")

## Check correlation between diapause and postdiapause counts:
cor.test(means_df$Diapause,
         means_df$Postdiapause)

## Remove low counts:
means_df_filtered <- subset(means_df,
                           Diapause > 10 &
                           Postdiapause > 10)

## Generate a plot using raw estmated counts for a random sample:
means_df_filtered_plot <- ggplot(data = means_df_filtered,
                             aes(x = Diapause,
                                 y = Postdiapause)) +
                        geom_point(alpha = 0.3, size = 4) +
                        xlab("Diapause mean raw gene-level counts") +
                        ylab("Postdiapause mean raw gene-level counts") +
                        geom_smooth(method = "lm", formula = y~x) +
                        theme_bw()

## Increase font size for text labels:
means_df_filtered_plot <- means_df_filtered_plot +
        theme(axis.title = element_text(face = "bold",
                                        size = "25"),
              axis.text.x = element_text(size = "17",
                                         angle = 45,
                                         hjust = 1),
              axis.text.y = element_text(size = "17"))

## Print to console:
means_df_filtered_plot

## Save to file:
ggsave(file = "correlation_plots_raw_estimated_counts.png",
       height = 10, width = 20)
```