--- 
title: "Bombus pesticide exposure: gene expression in head"
output: plot_DEGs_heatmap.html
author: Isabel Fletcher, Joe Colgan, Yannick Wurm http://wurmlab.com
--- 

## Introduction:

This script is for exploratory analysis of transcript quantification data and results of differential gene expression analysis.
Input files should be R objects saved and generated from previous scripts of DE analysis, including the object resulting from using tximport() on kallisto files, the significant transcript abundances selected from this object and a results table of significant genes, generated using DESeq2. 
This script will organise input data for generating heatmaps of significantly differentially expressed genes, based on statistical significance and log fold changes in expression. This script produces a heatmap that can be exported. 

```{r, message=FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2", "gplots", "heatmap3",
               "ggfortify", "rentrez",
               "reshape2", "scales", "ggpubr",
               "grid", "XML")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}

## Assign path for output:
output <- "results/"
dir.create(output, recursive = TRUE)
```

## Step One: Load files containing raw count estimates and differentially expressed genes:
At present this script only plots heatmaps for one pesticide and one control treatment. 
The script generates heatmap plots for two castes: workers and queens.

```{r, message = FALSE}
## Load raw counts generated by DESeq2:
load(file = "input/postdiapause_raw_counts.Rdata")

## Step Two: Organise data for generating a heatmap:
## Normalise counts:
normalised_counts_df <- t(scale(t(txi_counts$counts),
                                center = TRUE,
                                scale = TRUE))

## Update column names:
colnames(normalised_counts_df) <- c("infected_1",
                                    "infected_2",
                                    "infected_3",
                                    "infected_4",
                                    "infected_5",
                                    "infected_6",
                                    "infected_7",
                                    "control_1",
                                    "control_2",
                                    "control_3",
                                    "control_4",
                                    "control_5",
                                    "control_6",
                                    "control_7")

## Read in genes and annotations:
annotations <- read.table(file = "./input/gene_annotations_postdiapause.txt",
                          col.names = c("locus",
                                        "annotation"),
                          header = FALSE)

genelist <- as.character(unlist(annotations$locus))

annotations$annotation <- gsub("_", " ",
                               annotations$annotation)

## Select significant gene counts
normalised_counts_df_sig_df <- as.data.frame(normalised_counts_df[genelist, ])

## Step Three: Extract annotations for significant genes
## Using custom script, load gene annotations from NCBI for each
## significant gene:
gene_descriptions <- annotations$annotation

## Generate new sample names for plotting:
sample_names <- c(paste("infected", "_", 1:7, sep = ""),
                  paste("control", "_", 1:7, sep = ""))


## Extract columns of interest:
column_nums <- c(grep("infected",
                      colnames(normalised_counts_df_sig_df)),
                 grep("control",
                      colnames(normalised_counts_df_sig_df)))

## Subset columns of interest:
normalised_counts_subset <- normalised_counts_df_sig_df[, column_nums]

## Rename columns:
colnames(normalised_counts_subset) <- sample_names

## Add rownames with gene descriptions:
rownames(normalised_counts_subset) <- paste(annotations$locus, "-",
                                              annotations$annotation)

normalised_counts_subset$description <- paste(annotations$locus, "-",
                                              annotations$annotation)

## Step Four: Cluster samples by euclidean distance for plotting:
## Calculate eucleidean distance between rows:
dd <- hclust(dist(as.matrix(normalised_counts_subset[, 1:6])))

## Reorder based on cluster order:
normalised_counts_ordered <- as.data.frame(normalised_counts_subset[dd$order, ])

## Reshape ("melt") for plotting with ggplot2:
normalised_counts_ordered_melt <- melt(normalised_counts_ordered)

## Rename columns for plotting:
colnames(normalised_counts_ordered_melt) <- c("gene_name",
                                              "sample",
                                              "normalised_counts")

## Step Five: Cluster samples by euclidean distance for plotting:
## Reorder "gene_name" for plotting:
normalised_counts_ordered_melt$gene_name <- factor(normalised_counts_ordered_melt$gene_name,
               levels = rev(unique(c(as.character(unlist(normalised_counts_ordered_melt$gene_name))))))

## Create a treatment column to plot:
normalised_counts_ordered_melt$treatment <- rep(c("infected",
                                                  "control"),
                                                each = (nrow(normalised_counts_ordered_melt) / 2))

## Step Six: Generate a barplot:
pd_plot <- ggplot(normalised_counts_ordered_melt, aes(sample, gene_name)) +
        geom_tile(aes(fill = normalised_counts)) +
        geom_vline(xintercept = c(0.5, 7.5, 14.5)) +
        scale_fill_gradient2(low = "green",
                             mid = "black",
                             high = "red") +
        ylab("") +
                                xlab("") +
                                theme(legend.title = element_text(size = 12,
                                                                face = "bold"),
                                legend.text = element_text(size = 12,
                                                        face = "bold"),
                                axis.text.x = element_text(angle = 45,
                                                            size = 10,
                                                            hjust = 1,
                                                            face = "bold"),
                                axis.text.y = element_text(size = 12,
                                                            face = "plain",
                                                            colour = "black"),
                                axis.title = element_text(size = 12,
                                                           face = "bold"),
                                panel.grid.major = element_blank(),
                                panel.grid.minor = element_blank(),
                                panel.background = element_blank(),
                                legend.position = "top") +
                                scale_y_discrete(position = "right") +
                                labs(fill = "Log fold change") +
                                theme(axis.text.y = element_text(face = c(rep("bold", nrow(normalised_counts_ordered_melt)))))

## Save plot:
save(pd_plot,
     file = "./pd_plot.Rdata")

## Save output:
ggsave("./postdiapause_heatmap.png",
       width = 20,
       height = 18)
```

Run lintr

```{r, message = FALSE}
## Run lintr:
lintr::lint(file = "./plot_DEGs_heatmap.Rmd")
```


